File: python/wiringserial_smp01.py
import wiringpi
import sys
import time

def main():
    # initialize
    # wiringpi.wiringPiSetup()
    serial = wiringpi.serialOpen("/dev/ttyS5", 115200)
    if(serial < 0):
        print("Unable to open serial device.", file=sys.stderr)
        sys.exit(1)

    # main process
    wiringpi.serialPuts(serial, "\r\n")
    wiringpi.serialPuts(serial, "Start loop back.\r\n")
    while(True):
        data_avail = wiringpi.serialDataAvail(serial)
        if(data_avail > 0):
            character = wiringpi.serialGetchar(serial)
            if(character == 0x1b):
                break
            wiringpi.serialPutchar(serial, character)
        time.sleep(0.01)
    wiringpi.serialPuts(serial, "\r\n")
    wiringpi.serialPuts(serial, "End loop back.\r\n")
    wiringpi.serialClose(serial)

    sys.exit(0)

if(__name__ == "__main__"):
    main()

File: c_cpp/wiringserial_smp01/wiringserial_smp01.c
#include <stdio.h>
#include <wiringPi.h>
#include <wiringSerial.h>

int main(void) {
    /* initialize */
    // wiringPiSetup();
    int serial = serialOpen("/dev/ttyS5", 115200);
    if(serial < 0) {
        fprintf(stderr, "Unable to open serial device.\n");
        return 1;
    }

    /* main process */
    serialPuts(serial, "\r\n");
    serialPuts(serial, "Start loop back.\r\n");
    while(TRUE) {
        int data_avail = serialDataAvail(serial);
        if(data_avail > 0) {
            int character = serialGetchar(serial);
            if(character == 0x1b) {
                break;
            }
            serialPutchar(serial, character);
        }
        delay(10);
    }
    serialPuts(serial, "\r\n");
    serialPuts(serial, "End loop back.\r\n");
    serialClose(serial);

    return 0;
}
